sftp-app-starters-source-common/src/main/java/org/springframework/cloud/stream/app/sftp/common/source/SftpSourceRotator.java

```diff
@@ -48,93 +41,51 @@
  * @author David Turanski
  * @since 2.0
  */
-public class ListFilesRotator extends AbstractMessageSourceAdvice {
-
-	private static final Log logger = LogFactory.getLog(ListFilesRotator.class);
+public class SftpSourceRotator extends RotatingServerAdvice {
 
 	private final SftpSourceProperties properties;
+	private final StandardRotationPolicy rotationPolicy;
 
-	private final DelegatingSessionFactory<?> sessionFactory;
-
-	private final List<KeyDirectory> keyDirs = new ArrayList<>();
-
-	private final boolean fair;
-
-	private volatile boolean initialized;
-
-	private volatile Iterator<KeyDirectory> iterator;
-
-	private volatile KeyDirectory current;
-
-	public ListFilesRotator(SftpSourceProperties properties, DelegatingFactoryWrapper factory) {
+	public SftpSourceRotator(SftpSourceProperties properties, StandardRotationPolicy rotationPolicy) {
+		super(rotationPolicy);
 		this.properties = properties;
-		this.sessionFactory = factory.getFactory();
-		if (properties.isMultiSource()) {
-			this.keyDirs.addAll(SftpSourceProperties.keyDirectories(properties));
-		}
-		this.fair = properties.isFair();
-		this.iterator = this.keyDirs.iterator();
+		this.rotationPolicy = rotationPolicy;
 	}
 
 	public Map<String, Object> headers() {
 		Supplier<SftpSourceProperties.Factory> factory = () -> {
-			SftpSourceProperties.Factory selected = this.properties.getFactories().get(this.current.getKey());
+			SftpSourceProperties.Factory selected = this.properties.getFactories().get(this.getCurrentKey());
 			if (selected == null) {
 				// missing key used default factory
 				selected = this.properties.getFactory();
 			}
 			return selected;
 		};
 		Map<String, Object> map = new HashMap<>();
-		map.put(SFTP_SELECTED_SERVER_PROPERTY_KEY, new FunctionExpression<>(m -> this.current.getKey()));
+		map.put(SFTP_SELECTED_SERVER_PROPERTY_KEY, new FunctionExpression<>(m -> this.getCurrentKey()));
```

Comment:
Looks like all the entries are `FunctionExpression`. Why just don't make the whole `headers()` method to return a  `Map<String, Expression>` instead and avoid the mentioned in other place discussion about casting to the `FunctionExpression`